dace/sdfg/scope.py

```diff
@@ -0,0 +1,235 @@
+            # Add nodes of the next level to next_map_nodes
+            next_map_nodes.extend([
+                node for node in more_nodes if isinstance(node, nd.EntryNode)
+            ])
+        map_nodes = next_map_nodes
+
+    if include_entry:
+        children_nodes.add(entry_node)
+
+    # Preserve order of nodes
+    return ScopeSubgraphView(graph,
+                             [n for n in graph.nodes() if n in children_nodes],
+                             entry_node)
+
+
+def _scope_dict_inner(graph, node_queue, current_scope, node_to_children,
+                      result):
+    """ Returns a queue of nodes that are external to the current scope. """
+    # Initialize an empty list, if necessary
+    if node_to_children and current_scope not in result:
+        result[current_scope] = []
+
+    external_queue = collections.deque()
+
+    visited = set()
+    while len(node_queue) > 0:
+        node = node_queue.popleft()
+
+        # If this node has been visited already, skip it
+        if node in visited:
+            continue
+        visited.add(node)
+
+        # Set the node parent (or its parent's children)
+        if not node_to_children:
+            result[node] = current_scope
+        else:
+            result[current_scope].append(node)
+
+        successors = [n for n in graph.successors(node) if n not in visited]
+
+        # If this is an Entry Node, we need to recurse further
+        if isinstance(node, nd.EntryNode):
+            node_queue.extend(
+                _scope_dict_inner(graph, collections.deque(successors), node,
+                                  node_to_children, result))
+        # If this is an Exit Node, we push the successors to the external
+        # queue
+        elif isinstance(node, nd.ExitNode):
+            external_queue.extend(successors)
+        # Otherwise, it is a plain node, and we push its successors to the
+        # same queue
+        else:
+            node_queue.extend(successors)
+
+    return external_queue
+
+
+def _scope_dict_to_ids(state: 'dace.sdfg.SDFGState',
+                       scope_dict: ScopeDictType):
+    """ Return a JSON-serializable dictionary of a scope dictionary,
+        using integral node IDs instead of object references. """
+    def node_id_or_none(node):
+        if node is None: return -1
+        return state.node_id(node)
+
+    return {
+        node_id_or_none(k): [node_id_or_none(vi) for vi in v]
+        for k, v in scope_dict.items()
+    }
+
+
+def scope_contains_scope(sdict: ScopeDictType, node: NodeType,
+                         other_node: NodeType) -> bool:
+    """ 
+    Returns true iff scope of `node` contains the scope of  `other_node`.
+    """
+    curnode = other_node
+    nodescope = sdict[node]
+    while curnode is not None:
+        curnode = sdict[curnode]
+        if curnode == nodescope:
+            return True
+    return False
+
+
+def is_devicelevel(sdfg: 'dace.sdfg.SDFG', state: 'dace.sdfg.SDFGState',
```

Comment:
If this is for GPU only, make this clear from the function name

Reply:
I want it to be any device-level (the API existed before and now moved to this file)

Reply:
Fixed